src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -74,14 +74,20 @@ private static SocketAsyncEngine[] CreateEngines()
             return engines;
         }
 
+        /// <summary>
+        /// Each <see cref="SocketAsyncContext"/> is assigned a <see cref="GCHandle"/> while registered with a <see cref="SocketAsyncEngine"/>.
+        /// <para>The handle is used as the <see cref="Interop.Sys.SocketEvent.Data"/> to quickly map events to <see cref="SocketAsyncContext"/>s.</para>
+        /// <para>When a socket is disposed and unregistered, there is a small race condition where we may still receive events with the
+        /// now-unused <see cref="GCHandle"/>. Since we assume that the handle is always allocated in <see cref="SocketEventHandler.HandleSocketEvents"/>,
+        /// we can never <see cref="GCHandle.Free"/> these handles.</para>
+        /// <para>Instead, we reuse handles for new sockets. The race condition may therefore trigger a notification for the wrong socket,
+        /// but that is okay as those operations can be retried.</para>
+        /// </summary>
+        private static readonly ConcurrentQueue<GCHandle> s_gcHandles = new();
```

Comment:
Is this pool really necessary?

Reply:
It's not a typical pool. As the comment above describes, we're never freeing these handles since their lifetime is tricky.
This pool is here to avoid completely leaking a handle per socket, instead capping their number at whatever peak load the process reaches. That seems fine given that even with the dictionary, we wouldn't ever resize it back down.

Reply:
The dictionary has only a small overhead of memory per slot. What is the impact of never releasing a GCHandle?

Reply:
About 8 bytes * peak socket count

Reply:
@Maoni0?

Reply:
Not sure if this is related, but I've observed a lot of gchandle related events when I was profiling OrchardCMS benchmark:

<img width="680" alt="Screenshot 2024-10-20 at 18 18 34" src="https://github.com/user-attachments/assets/9c4ec7c0-9231-4b67-b3ea-cd59ae2dcbcf">